feat: Add PersonOverride model

tomasfarias opened the pull request:
## Problem

We need to store PersonOverrides to allow us to:
1. Join recent events with their correct Persons (if a merge or merge-like event occurred).
2. Propagate the overrides to the main events table to correct historical events.
<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

This model will store a 1:1 mapping of a Person to the Person it was overridden with in a merge or merge-like event.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

tomasfarias committed:
feat: Add PersonOverrides model

This model will store a 1:1 mapping of a Person to the Person it was
overridden with in a merge or merge-like event.

tomasfarias committed:
chore: Add migration file

tomasfarias committed:
fix: Update latest migration manifest

tiina303 commented on posthog/migrations/0291_create_person_override_model.py:
```diff
@@ -0,0 +1,31 @@
+        migrations.CreateModel(
+            name="PersonOverride",
+            fields=[
+                ("id", models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID")),
```
Should we use `BigAutoField` to not need to worry about running out of integer space? https://docs.djangoproject.com/en/4.1/ref/models/fields/#autofield

tomasfarias replied:
Given that in the worst case we could have 1 override for all persons, and persons table could reach the int32 limit, this is a good idea :+1:. Will add a commit.

tiina303 replied:
should this be in the models class somehow too?

tiina303 commented on posthog/models/person/person.py:
```diff
@@ -99,3 +99,24 @@ class Meta:
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["old_person_id", "override_person_id"], name="unique override per old_person_id"
```
shouldn't it be this (the team part isn't even really needed, but could help in the rare case of uuid collisions):
```suggestion
                fields=["team", "old_person_id"], name="unique override per old_person_id"
```

tomasfarias replied:
I think both conditions achieve the same thing. I'll commit this :+1:

tiina303 commented on posthog/models/person/person.py:
```diff
@@ -99,3 +99,24 @@ class Meta:
+    old_person_id = models.UUIDField(db_index=True)
+    override_person_id = models.UUIDField(db_index=True)
+
+    created_at: models.DateTimeField = models.DateTimeField(auto_now_add=True, blank=True)
```
this is the old_person created_at, correct? please add a comment that's not obvious. Also the motivation for this field to know whether this will be a short or long term override in CH.

tomasfarias replied:
Instead of a comment I could rename the field to `old_person_created_at` to dispel all doubts.

tomasfarias replied:
Also, I believe I missed `is_historical`/`is_longterm` flag we talked about...

tiina303 replied:
Yes, thought this through a bit more we'll want the flag. This would make it easier to sync from postgres to clickhouse if they got out of sync, as postgres is the source of truth here.

One could think that we can just use `old_person_created_at` value to figure out which table it should go to, but we can't because we don't know at which time the entry was added to postgres. Hence if we had a 45 day cut off and we see `old_person_created_at` that's 46 days ago it might have been added today (long term table) or it might have been added 2 days ago (short term table).

tomasfarias replied:
Do we need the flag then or a `override_added_at` field?

tomasfarias replied:
I'm pushing a commit to compute the flag in a second, fighting with some unit tests...

tomasfarias replied:
Ok, gave it a shot. The `is_long_term` flag requires `old_person_created_at` in order to compare whether that is past the cut off point from now. Let me know what you think.

tomasfarias replied:
We could also look up the `old_person_created_at` ourselves, thus not requiring that parameter. As the one who will be implementing the ingestion of these tables, what do you think?

neilkakkar commented on posthog/models/person/person.py:
```diff
@@ -99,3 +99,24 @@ class Meta:
     version: models.BigIntegerField = models.BigIntegerField(null=True, blank=True)
+
+
+class PersonOverride(models.Model):
```
I'm guessing there will be a similar table in CH for querying? and both of these stay in sync?

tiina303 replied:
yes - works very similarly to Person DistinctID. 

Why we need the table in postgres is explained here: https://docs.google.com/document/d/12Q1KcJ41TicIwySCfNJV5ZPKXWVtxT7pzpB3r9ivz_0/edit#bookmark=id.gpwgfub6blq9

Actually a bit more complex compared to distinct ID as we'll have 2 separate tables in CH (short and long term but exactly the same schema, we'll union them and then join for queries), read more: https://docs.google.com/document/d/12Q1KcJ41TicIwySCfNJV5ZPKXWVtxT7pzpB3r9ivz_0/edit#bookmark=id.9gse52dl0e4w

tomasfarias committed:
refactor: Use unique team to old_person_id constraint

Co-authored-by: Tiina Turban <tiina303@gmail.com>

tomasfarias committed:
fix: Use BigAutoField to avoid issues with int32 limit

tiina303 commented on posthog/migrations/0291_create_person_override_model.py:
```diff
@@ -0,0 +1,31 @@
+        migrations.AddConstraint(
+            model_name="personoverride",
+            constraint=models.UniqueConstraint(
+                fields=("old_person_id", "override_person_id"), name="unique override per old_person_id"
```
re-create the migration too

tomasfarias committed:
fix: Recreate migration with constraint and ID field

tomasfarias committed:
refactor: Update field name in PersonOverride

tomasfarias committed:
feat: Add is_long_term flag to PersonOverride

This flag computes whether the old_person that is being overriden was
created past the cutoff point (defaults to 45 days).

fuziontech reviewed (commented):
I just have a bit of a worry about how we are using two sets of logic in two spots to figure out whether or not a person override `is_long_term`

tomasfarias commented on posthog/models/person/person.py:
```diff
@@ -99,3 +100,39 @@ class Meta:
+class PersonOverrideManager(models.Manager):
+    """A manager for PersonOverride that sets the is_long_term flag on creation."""
+
+    def create_override(self, *args: Any, long_term_cutoff: int = 45, **kwargs: Any) -> "PersonOverride":
```
I'm not aware of how things will look (or already look) from the ingestion side. But if we already have some logic there to calculate `is_long_term` or if we plan to, then yeah I can drop this and just expose the flag for ingestion to set.

tomasfarias replied:
The alternative would be to remove the ingestion side of things and keep this. Again, I don't know how things look from that end so I can't make the call of which one to get rid of.

Maybe @tiina303 also has some input on this?

tomasfarias replied:
Seems like we'll just be using `old_person_created_at` (now renamed to `oldest_event`) and set that during ingestion. So the model need only expose the parameter. I've updated the PR :+1:

tiina303 replied:
Just to explain how it would have worked on ingestion side: we'd need to send the kafka events to one of the other table in CH, which would mean different kafka topics, so the ingestion code would need to know that. That said the code for figuring it if short or long term could have still lived here and we'd just get the row after we updated and based on what value we pick the kafka topic to use (this column/value could have even been the name of the kafka topic to make it simpler).

tomasfarias committed:
refactor: Use oldest_event instead of is_long_term flag

We decided to just use the oldest event date instead of a flag. This
value is to be set by ingestion, so the model only needs to expose it.

tomasfarias committed:
chore: Update migration

tiina303 reviewed (approved):
🚢

tomasfarias commented:
Had to rebase branch to master.

tomasfarias commented:
Code is exactly the same, so I can merge if everything's green.

tomasfarias committed:
fix: Correct type signature in excepthook override

tomasfarias commented:
Every check is red because of some rate limit error, but I think that's affecting all PRs :sweat_smile:
